api/users/users-router.js

```diff
@@ -1,44 +1,162 @@
+router.get("/", (req, res, next) => {
+  users
+    .get()
+    .then((users) => {
+      res.status(200).json(users);
+    })
+    .catch(next);
 });
 
-router.post('/', (req, res) => {
-  // RETURN THE NEWLY CREATED USER OBJECT
-  // this needs a middleware to check that the request body is valid
-});
+// RETURN THE USER OBJECT
+// this needs a middleware to verify user id
 
-router.put('/:id', (req, res) => {
-  // RETURN THE FRESHLY UPDATED USER OBJECT
-  // this needs a middleware to verify user id
-  // and another middleware to check that the request body is valid
+router.get("/:id", validateUserId(), (req, res, next) => {
+  res.status(200).json(req.user);
 });
 
-router.delete('/:id', (req, res) => {
-  // RETURN THE FRESHLY DELETED USER OBJECT
-  // this needs a middleware to verify user id
+// RETURN THE NEWLY CREATED USER OBJECT
+// this needs a middleware to check that the request body is valid
+
+router.post("/", validateUser(), (req, res, next) => {
+  users
+    .insert(req.body)
+    .then((newUser) => {
+      res.status(201).json(newUser);
+    })
+    .catch(next);
 });
 
-router.get('/:id/posts', (req, res) => {
-  // RETURN THE ARRAY OF USER POSTS
-  // this needs a middleware to verify user id
+// RETURN THE FRESHLY UPDATED USER OBJECT
+// this needs a middleware to verify user id
+// and another middleware to check that the request body is valid
+
+router.put(
+  "/:id",
+  validateUser(),
+  validateUserId(),
+  (req, res, next) => {
+    users
+      .update(req.params.id, req.body)
+      .then((updatedUser) => {
+        res.status(200).json(updatedUser);
+      })
+      .catch(next);
+  }
+);
+
+// RETURN THE FRESHLY DELETED USER OBJECT
+// this needs a middleware to verify user id
+
+router.delete("/:id", validateUserId(), (req, res, next) => {
+  users
+    .remove(req.params.id)
+    .then((deletedUser) => {
+      res.status(200).json(deletedUser);
+    })
+    .catch(next);
 });
 
-router.post('/:id/posts', (req, res) => {
-  // RETURN THE NEWLY CREATED USER POST
-  // this needs a middleware to verify user id
-  // and another middleware to check that the request body is valid
+// RETURN THE ARRAY OF USER POSTS
+// this needs a middleware to verify user id
+
+router.get("/:id/posts", validateUserId(), (req, res, next) => {
+  users
+    .getUserPosts(req.params.id)
+    .then((posts) => {
+      res.status(200).json(posts);
+    })
+    .catch(next);
 });
 
+// RETURN THE NEWLY CREATED USER POST
+// this needs a middleware to verify user id
+// and another middleware to check that the request body is valid
+
+router.post(
+  "/:id/posts",
+  validateUserId(),
+  validatePost(),
+  (req, res, next) => {
+    posts
+      .insert(req.body)
+      .then((newPost) => {
+        res.status(201).json(newPost);
+      })
+      .catch(next);
+  }
+);
+
+// RETURN THE POST OBJECT
+// this needs a middleware to verify user id
+// this needs a middleware to verify post id
+
+router.get(
```

Comment:
acest endpoint la tine e problematic asai? M-am uitat numai ce, e un bug tare nasty 😄 . Deci uite atent, in endpoint tu ce faci, returnezi post-ul gasit de middleware `validatePostId`, iar in acest middleware tu cauti post-ul dupa `req.params.id`, te rog sa verifici asta. Si daca te uiti atent cum este declarat path-ul tau `/:id/users/:postId`, atunci `req.params.id` ce o sa fie? Id-ul utilizatorului. Practic ce se intampla este ca tu cauti post-ul dupa id-ul utilizatorului in loc de cel al sau

Reply:
Ok, validatePostId  valideaza de fapt id-ul user-ului )). 
trebuie sa scriu acolo req.params.postId.